transport/http/src/main/java/org/kaazing/gateway/transport/http/PersistentConnectionPool.java

```diff
@@ -21,62 +21,90 @@
+
+    // server address -> set of persistent connections (per thread).
+    // Using ThreadLocal for the following reasons:
+    // - we need to return thread-aligned IoSession
+    // - it reduces contention compared to synchronized ServerConnections
+    private final ThreadLocal<ServerConnections> connections;
+
+    // No of connections per server across ALL threads
+    private final Map<HttpResourceAddress, AtomicInteger> counts;
+
 
-    // transport address -> set of persistent connections
-    private final Map<ResourceAddress, Set<IoSession>> connections;
     private final Logger logger;
-    private final CloseListener closeListener;
     private final HttpConnectIdleFilter idleFilter;
+    private final CloseListener closeListener;
 
     PersistentConnectionPool(Logger logger) {
-        this.connections = new HashMap<>();
+        this.connections = new VicariousThreadLocal<ServerConnections>() {
+            @Override
+            protected  ServerConnections initialValue() {
+                return new ServerConnections();
+            }
+        };
+        this.counts = new ConcurrentHashMap<>();
         this.logger = logger;
+        this.idleFilter = new HttpConnectIdleFilter(logger);
         this.closeListener = new CloseListener(this);
-        this.idleFilter = new HttpConnectIdleFilter(this, logger);
     }
 
     /*
-     * Recycle existing transport session so that it can be used as a http
-     * persistent connection
+     * Cache existing transport session so that it can be used for future
+     * connect requests to server
+     *
+     * @return true if the idle connection is cached for reuse
+     *         false otherwise
      */
-    public void recycle(IoSession session, Integer keepAliveTimeout) {
-        assert keepAliveTimeout != null;
-        if (logger.isDebugEnabled()) {
-            logger.debug(String.format("Recycling (adding to pool) http connect persistent connection %s", session));
+    boolean recycle(DefaultHttpSession httpSession) {
+        if (!add(httpSession)) {
+            return false;
         }
-        ResourceAddress address = BridgeSession.REMOTE_ADDRESS.get(session);
-        Set<IoSession> sessions = connections.get(address);
-        if (sessions == null) {
-            sessions = new HashSet<>();
-            connections.put(address, sessions);
-        }
-        sessions.add(session);
-        CloseFuture closeFuture = session.getCloseFuture();
+
+        HttpResourceAddress serverAddress = (HttpResourceAddress)httpSession.getRemoteAddress();
+        IoSession transportSession = httpSession.getParent();
+
+        SERVER_ADDRESS.set(transportSession, serverAddress);
+
+        // Take care of transport session close
+        CloseFuture closeFuture = transportSession.getCloseFuture();
         closeFuture.addListener(closeListener);
-        addIdleFilter(session, keepAliveTimeout);
+
+        // Track transport session idle
```

Comment:
Perhaps you can elaborate in this comment to describe the requirements for tracking session idle?  Restating what the code is doing in some comment is typically a cause to remove said comment, but giving motivations / requirements is useful for future reading of the code to understand why some code is there.